exporter/signalfxexporter/translation/converter.go

```diff
@@ -475,5 +467,113 @@ func (c *MetricsConverter) ConvertDimension(dim string) string {
+// datapoints with metric name greater than 256 characters.
+func (dpv *datapointValidator) sanitizeDataPoints(dps []*sfxpb.DataPoint) []*sfxpb.DataPoint {
+	resultDatapointsLen := 0
+	for dpIndex, dp := range dps {
+		if dpv.isValidMetricName(dp.Metric) {
+			dp.Dimensions = dpv.sanitizeDimensions(dp.Dimensions)
+			if resultDatapointsLen < dpIndex {
+				dps[resultDatapointsLen] = dp
+			}
+			resultDatapointsLen++
+		}
+	}
+
+	// Trim datapoints slice to account for any removed datapoints.
+	return dps[:resultDatapointsLen]
+}
+
+// sanitizeDimensions replaces all characters unsupported by SignalFx backend
+// in metric label keys and with "_" and drops dimensions when the key is greater
+// than 128 characters or when value is greater than 256 characters in length.
+func (dpv *datapointValidator) sanitizeDimensions(dimensions []*sfxpb.Dimension) []*sfxpb.Dimension {
+	resultDimensionsLen := 0
+	for dimensionIndex, d := range dimensions {
+		if dpv.isValidDimension(d) {
+			d.Key = filterKeyChars(d.Key, dpv.nonAlphanumericDimChars)
+			if resultDimensionsLen < dimensionIndex {
+				dimensions[resultDimensionsLen] = d
+			}
+			resultDimensionsLen++
+		}
+	}
+
+	// Trim dimensions slice to account for any removed dimensions.
+	return dimensions[:resultDimensionsLen]
+}
+
+func (dpv *datapointValidator) isValidMetricName(name string) bool {
+	if len(name) > maxMetricNameLength {
+		dpv.logger.Warn("dropping datapoint",
+			zap.String("reason", "metric name too long"),
+			zap.String("metric_name", name))
+		return false
+	}
+	return true
+}
+
+func (dpv *datapointValidator) isValidDimension(dimension *sfxpb.Dimension) bool {
+	return dpv.isValidDimensionName(dimension.Key) && dpv.isValidDimensionValue(dimension.Value)
+}
+
+func (dpv *datapointValidator) isValidDimensionName(name string) bool {
+	if len(name) > maxDimensionNameLength {
+		dpv.logger.Warn("dropping dimension",
+			zap.String("reason", "dimension name too long"),
+			zap.String("dimension_name", name))
+		return false
+	}
+	return true
+}
+
+func (dpv *datapointValidator) isValidDimensionValue(value string) bool {
+	if len(value) > maxDimensionValueLength {
+		dpv.logger.Warn("dropping dimension",
+			zap.String("reason", "dimension value too long"),
+			zap.String("dimension_value", value))
+		return false
+	}
+	return true
+}
+
+// Copied from https://github.com/open-telemetry/opentelemetry-collector/blob/v0.26.0/exporter/exporterhelper/queued_retry.go#L108
+func createSampledLogger(logger *zap.Logger) *zap.Logger {
+	if logger.Core().Enabled(zapcore.DebugLevel) {
+		// Debugging is enabled. Don't do any sampling.
+		return logger
+	}
+
+	// Create a logger that samples all messages to 1 per 10 seconds initially,
+	// and 1/100 of messages after that.
+	opts := zap.WrapCore(func(core zapcore.Core) zapcore.Core {
+		return zapcore.NewSamplerWithOptions(
+			core,
+			10*time.Second,
+			1,
+			100,
```

Comment:
Let's go with 1/10000 here. This is potentially a very high rate of logs since it is per data point, so 1/100 may not be enough to avoid log flooding.

Reply:
Updated.